File: src/lib/logic/shots.ts
/**
 * Pure logic behind the screenshot offer. tech.md 6.13.
 */

import type { ShotOffer } from '$lib/types/generated/ShotOffer';

/**
 * How much of the offer is left, from 1 at the moment it went up to 0 when it
 * is over. Total over any clock: an offer already gone reads as 0 rather than
 * as a negative bar, and one from the future reads as full.
 */
export function timeLeft(offer: ShotOffer, now: number): number {
  const span = offer.expires_at - offer.created_at;
  if (!Number.isFinite(span) || span <= 0) return 0;
  const left = (offer.expires_at - now) / span;
  if (Number.isNaN(left)) return 0;
  return Math.min(1, Math.max(0, left));
}

/**
 * Whole seconds left on the offer, rounded up, and never below zero.
 *
 * Up rather than down: with 4.2 seconds left the honest answer to "have I got
 * time to reach the key" is five, and a `0s` standing on screen while the key
 * still works reads as a broken offer. Zero is never shown, because at zero
 * there is no offer left to show it on. tech.md 6.13.
 */
export function secondsLeft(offer: ShotOffer, now: number): number {
  const left = offer.expires_at - now;
  if (!Number.isFinite(left) || left <= 0) return 0;
  return Math.ceil(left / 1000);
}

/**
 * Whether a paste is a picture rather than text, judged before anything reads
 * the clipboard's contents. tech.md 6.13.
 *
 * The webview pastes text itself and must be left to it: intercepting every
 * paste to ask Rust would put a system paste prompt in front of somebody
 * pasting a word. So text wins whenever it is there, and Rust is asked only
 * when the types say picture and nothing says text. Reading the types raises
 * nothing; reading the contents is what does.
 */
export function looksLikeImagePaste(types: readonly string[]): boolean {
  if (types.some((type) => type === 'text/plain' || type === 'text/html')) return false;
  return types.some((type) => type.startsWith('image/') || type === 'Files');
}

/**
 * A line of a reply that is an attached screenshot, and not prose that
 * happens to mention one. tech.md 6.13.
 *
 * By shape and never by reading the disk: the whole line, the shots
 * directory, a ulid, `.png`. A path Peekle wrote is the only thing that
 * matches, so a person quoting a filename keeps their words.
 */
const SHOT_LINE = /^\/.*\/peekle\/shots\/[0-9A-HJKMNP-TV-Z]{26}\.png$/;

/**
 * Whether this path is one Peekle wrote for a screenshot.
 *
 * By shape, never by reading the disk, and used by more than the split below:
 * the chip above the field asks it to know whether it has a picture to show,
 * and the file rule (6.25) asks it to keep its hands off a screenshot.
 */
export function isShot(path: string): boolean {
  return SHOT_LINE.test(path.trim());
}

/** A reply split into the shots it carries and the words that go with them. */
export interface SaidWithShots {
  shots: string[];
  said: string;
}

/**
 * What a reply says and what it carries. `compose` puts each path on its own
 * line before the text, so the split is the same one in reverse. tech.md 6.13.
 */
export function shotLines(text: string): SaidWithShots {
  const shots: string[] = [];
  const words: string[] = [];

  for (const line of text.split('\n')) {
    if (isShot(line)) {
      shots.push(line.trim());
      continue;
    }
    words.push(line);
  }

  return { shots, said: words.join('\n').trim() };
}

/**
 * How big the picture is, in the words the CLI uses for the same thing:
 * `1172×246`. tech.md 6.13.
 *
 * `null` until the picture has been measured, which is what a browser answers
 * with before it has loaded one: a size of `0×0` on screen is worse than no
 * size at all. Read off the image itself rather than carried on the entry --
 * the webview has already loaded the file to draw it, and a number Rust sent
 * would be a second answer to a question the picture answers.
 */
export function shotSize(width: number, height: number): string | null {
  if (!Number.isFinite(width) || !Number.isFinite(height)) return null;
  if (width <= 0 || height <= 0) return null;
  return `${Math.round(width)}\u00d7${Math.round(height)}`;
}

/**
 * What the chip above the field calls the attachment. The full path is what
 * the agent gets, and it is far too long to sit over a reply box.
 */
export function shotName(path: string): string {
  const name = path.split('/').filter(Boolean).pop();
  return name ?? path;
}

File: src-tauri/src/shots.rs
//! The screenshot watch. tech.md 6.13.
//!
//! Control-Shift-Command-4 puts an image on the pasteboard and nowhere else,
//! and the channel to an agent carries text. So Peekle watches for the write,
//! offers to attach it, and on agreement turns the image into a file and the
//! file into a line of the next reply.
//!
//! AppKit lives here and in `panel.rs`, and nowhere else. tech.md 12.
//!
//! Two halves, deliberately unequal. Detection reads the change count and the
//! type names, which raises nothing and copies nothing, and it runs on a
//! timer. Reading the contents runs once, after the user pressed the key: from
//! macOS 15 that read is a system paste prompt, and asking for one on every
//! copy anybody makes would be a product that spies. tech.md R-13.

use std::sync::Arc;
use std::time::Duration;

use peekle_core::shots::{self, Pasteboard};
use peekle_core::types::{IslandView, ShotOffer, ToastRequest, ToastTone};
use tauri::{AppHandle, Emitter, Manager};

use crate::events;
use crate::hotkey;
use crate::panel;
use crate::state::AppState;
use crate::windows;

/// The real pasteboard.
pub struct SystemPasteboard;

impl Pasteboard for SystemPasteboard {
    fn change_count(&self) -> i64 {
        use objc2_app_kit::NSPasteboard;

        let pasteboard = NSPasteboard::generalPasteboard();
        pasteboard.changeCount() as i64
    }

    fn item_types(&self) -> Vec<Vec<String>> {
        use objc2_app_kit::NSPasteboard;

        let pasteboard = NSPasteboard::generalPasteboard();
        // The item and not the pasteboard: NSPasteboard synthesises TIFF from
        // a PNG, so its declared list calls every screenshot an image and
        // every image a screenshot. tech.md 6.13.
        let Some(items) = pasteboard.pasteboardItems() else {
            return Vec::new();
        };
        items
            .iter()
            .map(|item| item.types().iter().map(|t| t.to_string()).collect())
            .collect()
    }

    fn read_png(&self) -> Option<Vec<u8>> {
        use objc2_app_kit::{NSPasteboard, NSPasteboardTypePNG};

        let pasteboard = NSPasteboard::generalPasteboard();
        // The extern static is the type name itself, and reading one is what
        // `unsafe` covers here. Nothing about the read is fallible otherwise.
        let png = unsafe { NSPasteboardTypePNG };
        let data = pasteboard.dataForType(png)?;
        Some(data.to_vec())
    }
}

/// Starts the watch. Silent and cheap: one counter read per tick, and the
/// contents are never touched here. tech.md 6.13.
pub fn watch(app: &AppHandle, state: Arc<AppState>) {
    let (enabled, poll_ms, spelling) = {
        let config = state.lock_config();
        (
            config.shots.enabled,
            config.shots.poll_ms,
            config.hotkey.attach.clone(),
        )
    };
    if !enabled {
        tracing::debug!("the screenshot watch is switched off");
        return;
    }
    if spelling.is_empty() {
        // Nothing could answer an offer, and an offer nobody can answer is
        // just a line across the screen. tech.md 6.13.
        tracing::debug!("no attach key configured, so no screenshot offers");
        return;
    }

    // From what is on the pasteboard now, not from zero: the first tick must
    // not offer to attach whatever the user copied before Peekle started.
    state.seed_change_count(state.pasteboard.change_count());

    let handle = app.clone();
    tauri::async_runtime::spawn(async move {
        let mut ticker = tokio::time::interval(Duration::from_millis(poll_ms));
        loop {
            ticker.tick().await;
            tick(&handle, &state);
        }
    });
}

fn tick(app: &AppHandle, state: &Arc<AppState>) {
    let now = now_ms();

    // An offer nobody answered settles itself, and the Up arrow goes back to
    // whatever application the user is actually looking at. tech.md 6.13.
    if let Some(offer) = state.shot.take_expired(now) {
        tracing::debug!(id = %offer.id, "nobody answered the screenshot offer");
        close_offer(app, state);
        put_away(app, state);
    }

    let count = state.pasteboard.change_count();
    if !state.pasteboard_changed(count) {
        return;
    }
    if !shots::is_screenshot(&state.pasteboard.item_types()) {
        return;
    }
    open_offer(app, state, now);
}

/// Puts the offer up, or decides there is nothing to offer.
fn open_offer(app: &AppHandle, state: &Arc<AppState>, now: i64) {
    // An active permission request already holds the view and is the
    // stronger claim, the same way it outranks a `Stop` reveal. tech.md 6.7.
    // Forcing the pill over it would not queue behind it: `set_view` simply
    // overwrites the view, so the permission's own `Session` view is gone and
    // nothing ever brings it back once the pill's five seconds are up --
    // the pending hook is still blocking, but there is no longer any UI left
    // that can answer it. That is the freeze this guards against.
    if state.active_prompt().is_some() {
        tracing::debug!("a permission request holds the view, so no screenshot offer");
        return;
    }

    let Some(card) = state.newest_owned_session() else {
        // A screenshot with nowhere to go. An observed session has no input
        // field, so there is nothing to attach it to. tech.md 6.5 and 6.13.
        tracing::debug!("a screenshot arrived with no session of ours to take it");
        return;
    };

    let (secs, spelling) = {
        let config = state.lock_config();
        (config.shots.offer_secs, config.hotkey.attach.clone())
    };

    let offer = ShotOffer {
        id: shots::new_id(),
        session_id: card.session.session_id.clone(),
        project: card.session.project.clone(),
        created_at: now,
        expires_at: now + i64::from(secs) * 1000,
    };

    // The fresh screenshot pushes the stale one out: one key cannot answer for
    // two, and the one the user wants is the one they just took.
    if let Some(replaced) = state.shot.open(offer.clone()) {
        tracing::debug!(id = %replaced.id, "a newer screenshot replaced the offer");
    }

    if let Err(err) = hold_attach_key(app, state, &spelling) {
        // Nothing could answer it, so nothing is shown. Better than a line on
        // screen naming a key that does nothing. tech.md 6.13 and R-14.
        tracing::warn!(error = %err, "the attach key is unavailable, so no offer");
        state.shot.take();
        return;
    }

    tracing::debug!(session = %offer.session_id, "offering to attach a screenshot");
    emit_offer(app, Some(&offer));
    // The offer is one line of status, which is what `Pill` is. It runs on its
    // own clock exactly like a toast does. tech.md 6.7.
    windows::set_view(app, IslandView::Pill);
}

/// The user agreed. The one path that reads the contents of the pasteboard.
pub fn attach(app: &AppHandle) {
    let state = app.state::<Arc<AppState>>().inner().clone();

    // Exactly once. Everything below hangs off this take, so an offer that
    // expired a moment ago writes no file and opens no window. tech.md 6.13.
    let Some(offer) = state.shot.take() else {
        tracing::debug!("the attach key fired with no offer standing");
        return;
    };
    close_offer(app, &state);

    let Some(png) = state.pasteboard.read_png() else {
        // Every path from here down leaves the island where the offer left it,
        // so the pill has to come down on its own.
        // The user copied something else between the offer and the answer.
        tracing::warn!("the screenshot left the pasteboard before it was attached");
        say(app, "The screenshot is gone");
        return;
    };

    let Some(dir) = shots::shots_dir() else {
        tracing::warn!("no cache directory for screenshots");
        say(app, "Nowhere to save the screenshot");
        return;
    };
    let keep = state.lock_config().shots.keep;

    let path = match shots::write_shot(&dir, &offer.id, &png, keep) {
        Ok(path) => path.to_string_lossy().to_string(),
        Err(err) => {
            tracing::warn!(error = %err, "could not save the screenshot");
            say(app, "Could not save the screenshot");
            return;
        }
    };

    tracing::info!(session = %offer.session_id, bytes = png.len(), "attached a screenshot");
    let payload = serde_json::json!({ "session_id": offer.session_id, "path": path });
    if let Err(err) = app.emit_to(panel::ISLAND, events::SHOT_ATTACHED, payload) {
        tracing::warn!(error = %err, "failed to emit shot-attached");
    }

    // Attaching is not sending, so the island opens on the session with the
    // shot waiting in the field and the user writes what they want said.
    // tech.md 6.13.
    let handle = app.clone();
    let session_id = offer.session_id.clone();
    tauri::async_runtime::spawn(async move {
        windows::reveal_turn(&handle, &session_id).await;
    });
}

/// Drops the key and tells the island the offer is over. Runs on every path
/// that settles one, because a leaked offer holds the Up arrow away from the
/// whole system. tech.md 6.13 and R-14.
fn close_offer(app: &AppHandle, state: &Arc<AppState>) {
    release_attach_key(app, state);
    emit_offer(app, None);
}

/// Takes the pill down, and only the pill. Something more important may have
/// opened while the offer stood, and collapsing then would throw away a
/// session the user is reading. The agreement path skips this entirely: it
/// opens the session next, and a collapse in between reads as a glitch.
fn put_away(app: &AppHandle, state: &Arc<AppState>) {
    if state.view() == IslandView::Pill {
        windows::set_view(app, IslandView::Collapsed);
    }
}

fn emit_offer(app: &AppHandle, offer: Option<&ShotOffer>) {
    let payload = serde_json::json!({ "offer": offer });
    if let Err(err) = app.emit_to(panel::ISLAND, events::SHOT, payload) {
        tracing::warn!(error = %err, "failed to emit the screenshot offer");
    }
}

fn hold_attach_key(
    app: &AppHandle,
    state: &Arc<AppState>,
    spelling: &str,
) -> Result<(), peekle_hotkey::HotkeyError> {
    if !state.set_attach_key(true) {
        // Already held, from an offer this one replaced.
        return Ok(());
    }
    match hotkey::register(app, spelling) {
        Ok(()) => Ok(()),
        Err(err) => {
            state.set_attach_key(false);
            Err(err)
        }
    }
}

fn release_attach_key(app: &AppHandle, state: &Arc<AppState>) {
    if !state.set_attach_key(false) {
        return;
    }
    let spelling = state.lock_config().hotkey.attach.clone();
    hotkey::unregister(app, &spelling);
}

/// One line in the notch. Used only where the user pressed a key and got
/// nothing: silence there reads as a broken key rather than a missing shot.
fn say(app: &AppHandle, text: &str) {
    windows::toast(
        app,
        ToastRequest {
            text: text.to_string(),
            tone: ToastTone::Warn,
            ttl_ms: 1600,
            badge: None,
        },
    );
}

fn now_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or_default()
}
